sdk/core/src/bytes_response.rs

```diff
@@ -0,0 +1,138 @@
+#[cfg(feature = "mock_transport_generate")]
+use crate::{collect_pinned_stream, BytesStream, Response, StreamError};
+use bytes::Bytes;
+use http::{HeaderMap, StatusCode};
+use serde::de::Visitor;
+use serde::ser::{Serialize, SerializeStruct, Serializer};
+use serde::{Deserialize, Deserializer};
+use std::collections::HashMap;
+
+const FIELDS: &'static [&'static str] = &["status", "headers", "body"];
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct BytesResponse {
+    status: StatusCode,
+    headers: HeaderMap,
+    body: Bytes,
+}
+
+impl Serialize for BytesResponse {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut hm = HashMap::new();
+        for (h, v) in self.headers.iter() {
+            hm.insert(h.to_string(), v.to_str().unwrap());
+        }
+
+        let mut state = serializer.serialize_struct("Response", 3)?;
+        state.serialize_field(FIELDS[0], &self.status.as_u16())?;
+        state.serialize_field(FIELDS[1], &hm)?;
+        state.serialize_field(FIELDS[2], &self.body as &[u8])?;
+
+        state.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for BytesResponse {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_struct("Response", FIELDS, ByteResponseVisitor)
+    }
+}
+
+struct ByteResponseVisitor;
+
+impl<'de> Visitor<'de> for ByteResponseVisitor {
+    type Value = BytesResponse;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        formatter.write_str("struct BytesResponse")
+    }
+
+    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::MapAccess<'de>,
+    {
+        let status: (&str, u16) = match map.next_entry()? {
+            Some((a, b)) => (a, b),
+            None => return Err(serde::de::Error::custom("missing status code")),
+        };
+
+        if status.0 != FIELDS[0] {
+            return Err(serde::de::Error::custom(format!(
+                "unexpected field {}, expected {}",
+                status.0, FIELDS[0]
+            )));
+        }
+
+        let headers: (&str, HashMap<&str, String>) = match map.next_entry()? {
```

Comment:
As mentioned in the meeting, at some point we *must* sanitize headers. We do that in the SDKs separately - but often relatedly - for logging (all headers sanitized by default, but standard headers opted in in azure-core + more that the client defines are safe) and for test serialization. For example, in .NET we have an extensible sanitizer.

Given the increased security as of late for GitHub and internal services, I think we need a sanitizer now (we can sort out the logging later). Here's .NET's, for example, that is added to the test recorder by default (we can pass our own derivative in for customization instead): https://github.com/Azure/azure-sdk-for-net/blob/18ed528ddba7581418d42064a3c5e9dceb0e0833/sdk/core/Azure.Core.TestFramework/src/RecordedTestSanitizer.cs

Reply:
+1 for the folder name, I've changed it to `SessionRecords` to match the other SDKs.